### PR TITLE
While loop issue (was Spell checking)

### PR DESCRIPTION
**Pull request by ciscoheat**
I've started a spell checker, based on a use case from a classic forum discussion, if you don't mind... :)

It's not tested to actually work (apologies for bugs) and probably hasn't a very good design yet (advice very welcome), but I've run into some problems that I'd like to share. Calling RoleMethods won't work for some reason.

### CODE REVIEW

**Comment by ciscoheat** [comment]
Digging deeper into this, it seems that accessing `currentPos_` from within a RoleMethod is causing the problem, for example in `SpellCheck::Text::isFinished`.

File: tests/spell_check.k [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        // Sorry, English only...
+        private boolean isAlphabetic(String character) const {
+            assert(character.length() == 1);
+            return character == "-" || 
+                   (character >= "a" && character <= "z") ||
+                   (character >= "A" && character <= "Z");
+        }
+
+        private String nextCharacter() {
+            // NOTE: Will this postfix "trick" work? Untested...
```
Comment (ciscoheat) [inline tests/spell_check.k]: I mean prefix of course. :)

File: tests/spell_check.k [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+
+    public String getText() { return Buffer.getText(); }
+}
+
+///////////////////////////////////////
+
+context SpellCheck {
+    role Text {
+        public boolean isFinished() const {
+            return currentPos_ >= getText().length();
```
Comment (ciscoheat) [inline tests/spell_check.k]: Accessing `currentPos_` inside a RoleMethod crashes with this error:

```
Exception in thread "AWT-EventQueue-0" java.lang.AssertionError
    at info.fulloo.trygve.run_time.RTObjectCommon$RTContextObject.getRoleOrStagePropBinding(RTObjectCommon.java:223)
    at info.fulloo.trygve.run_time.RTExpression$RTIdentifier.run(RTExpression.java:1325)
    at info.fulloo.trygve.run_time.RunTimeEnvironment.runner(RunTimeEnvironment.java:362)
    at info.fulloo.trygve.run_time.RTExpression$RTMessage.pushArgumentLoop(RTExpression.java:786)
    at info.fulloo.trygve.run_time.RTExpression$RTMessage.run(RTExpression.java:918)
    at info.fulloo.trygve.run_time.RunTimeEnvironment.runner(RunTimeEnvironment.java:362)
    at info.fulloo.trygve.run_time.RTExpression$RTMessage.pushArgumentLoop(RTExpression.java:786)
    at info.fulloo.trygve.run_time.RTExpression$RTMessage.run(RTExpression.java:918)
    at info.fulloo.trygve.run_time.RunTimeEnvironment.runner(RunTimeEnvironment.java:362)
    at info.fulloo.trygve.run_time.RunTimeEnvironment.run(RunTimeEnvironment.java:153)
    at info.fulloo.trygve.editor.TextEditorGUI.runButtonActionPerformed(TextEditorGUI.java:590)
    at info.fulloo.trygve.editor.TextEditorGUI$7.actionPerformed(TextEditorGUI.java:216)
    at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
    at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2346)
    at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
    at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
    at javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:252)
    ...
```

**Comment by ciscoheat** [comment]
Even when using a parameter instead of `currentPos_`, like: 

``` java
private boolean isFinished(int pos) {
    return pos >= getText().length();
}
```

`isFinished` still cannot be called. `pos >= getText().length()` works in `nextWord` however.

**Comment by jcoplien** [comment]
Hej, Andreas,

I have a handle on the problem. It goes pretty deep. I have it half fixed but, because I’m teaching this week. I also have the \n feature implemented but not yet pushed to GitHub.

I’ll keep you posted.

**Comment by ciscoheat** [comment]
Thank you, no rush! :)

**Comment by jcoplien** [comment]
Give it a try. I’ve fixed a very, very nasty bug where it was getting confused about the Context playing the Role and Context containing the Role — a very intricate and difficult fix. Thanks for reporting this.

It also surfaced another good discussion issue for the list about whether a Role method should be able to access the symbols of its enclosing Context… You’ll see the warnings...

This could be a really canonical, great, accessible example once it’s polished up!!!

**Comment by ciscoheat** [comment]
It's working! With that, and the new warnings, I managed to finish the parser. Some questions and comments:
1. I used the Context for the role `SpellChecker`, according to the new warning. The RoleObjectContract for SpellChecker became a character position marker, basically. The `Text` role became simpler, but I realized Text isn't a part of the use case. Is that a problem?
2. I noticed that private methods of the Context (`currentPosition`, `advancePosition`) was ok to use when the Context played a Role inside itself. Seems like a nice feature, just want to check if it's intentional.
3. A more low-level question: Right now I'm calling `SpellChecker.advancePosition` after every parser step. I don't want to introduce any side-effects in `Text.currentCharacter` and advance it automatically there, so is this verbosity fine? I'm trying to keep the parser code simple, but parsing always feels a bit clunky.

**Comment by ciscoheat** [comment]
The current version 364e6aa gives a very strange behavior. After executing the line `if(suggestions.size() == 0)` (line 171) in `SpellCheck.Dictionary.checkSpelling`, the while loop in `SpellCheck.test` starts looping infinitely, or at least until the safety is released! If you comment out that if statement, it works.

**Comment by ciscoheat** [comment]
Just tested this against d48a2c648a, the problem still persists. It actually works when inverting the if statement:

``` java
// Moving the join string into the if statement
// instead of returning makes it work...!
if(suggestions.size() > 0) {
    String joined = "";
    int end = suggestions.size()-1;
    for(int i = 0; i <= end; i++) {
        joined = joined + suggestions.get(i);
        if(i < end) joined = joined + ", ";
    }
    System.out.println(word + " -> [" + joined + "]");
}
```

**Comment by ciscoheat** [comment]
Thanks for the "Context script may enact only Role scripts" error, it gives a good perspective on the difference between interfaces and roles. It makes this example outdated, but the strange while loop behavior is still there, so I have cleaned up the example a little. If you look at the `Dictionary` RoleMethods, I've added a few comments.

**Comment by ciscoheat** [comment]
It works for me, so I'll close this. Thanks for the fix!

**Pull request closed without merging**